### PR TITLE
[SEDONA-113] Add ST_PointN to Apache Sedona

### PR DESCRIPTION
**Pull request by Pooja444**
## Did you read the Contributor Guide?

- Yes, I have read [Contributor Rules](https://sedona.apache.org/community/rule/) and [Contributor Development Guide](https://sedona.apache.org/community/develop/)

## Is this PR related to a JIRA ticket?

- Yes, the URL of the assoicated JIRA ticket is https://issues.apache.org/jira/browse/SEDONA-113. The PR name follows the format `[SEDONA-XXX] my subject`.

## What changes were proposed in this PR?

Added ST_PointN function to Flink and SQL

## How was this patch tested?

Added unit tests in Java, Scala and Python

## Did this PR include necessary documentation updates?

- Yes, I have updated the documentation update.

### CODE REVIEW

File: core/src/main/java/org/apache/sedona/core/utils/GeomUtils.java [hunk 1]
```diff
@@ -81,6 +84,38 @@ public static Geometry getInteriorPoint(Geometry geometry) {
...
+        if (n > 0) {
+            nthCoordinate[0] = lineString.getCoordinates()[n - 1];
+        } else {
+            nthCoordinate[0] = lineString.getCoordinates()[p + n];
+        }
+        return new Point(new CoordinateArraySequence(nthCoordinate), lineString.getFactory());
+    }
+
+    public static Geometry getExteriorRing(Geometry geometry) {
+        return geometry.getFactory().createLinearRing(geometry.getCoordinates());
```
Comment (jiayuasu) [inline core/src/main/java/org/apache/sedona/core/utils/GeomUtils.java:116]: Use this function in JTS Polygon. Dont create LinearRing by coordinates as they might NOT be the rings.https://locationtech.github.io/jts/javadoc/org/locationtech/jts/geom/Polygon.html#getExteriorRing--

File: docs/api/flink/Function.md [hunk 2]
```diff
@@ -152,6 +152,34 @@ Input: `POLYGON ((-0.5 -0.5, -0.5 0.5, 0.5 0.5, 0.5 -0.5, -0.5 -0.5))`
 
 Output: `POLYGON ((-0.5 -0.5, 0.5 -0.5, 0.5 0.5, -0.5 0.5, -0.5 -0.5))`
 
+## ST_PointN
```
Comment (jiayuasu) [inline docs/api/flink/Function.md:155]: Please add docs for ST_ExteriorRing too.

File: sql/src/main/scala/org/apache/sedona/sql/UDF/Catalog.scala [hunk 3]
```diff
@@ -105,6 +105,7 @@ object Catalog {
     ST_Multi,
     ST_PointOnSurface,
     ST_Reverse,
+    ST_PointN,
```
Comment (jiayuasu) [inline sql/src/main/scala/org/apache/sedona/sql/UDF/Catalog.scala:108]: Register ST_ExteriorRing as well

**Review by jiayuasu — CHANGES_REQUESTED** [review]

File: sql/src/main/scala/org/apache/sedona/sql/UDF/Catalog.scala (hunk 3 above)
Comment (Pooja444) [reply sql/src/main/scala/org/apache/sedona/sql/UDF/Catalog.scala:108]: @jiayuasu It was already implemented in Scala, so it's already present in the file

File: docs/api/flink/Function.md (hunk 2 above)
Comment (Pooja444) [reply docs/api/flink/Function.md:155]: Added

File: core/src/main/java/org/apache/sedona/core/utils/GeomUtils.java (hunk 1 above)
Comment (Pooja444) [reply core/src/main/java/org/apache/sedona/core/utils/GeomUtils.java]: Sure, I updated this

File: docs/api/flink/Function.md [hunk 4]
```diff
@@ -152,6 +152,55 @@ Input: `POLYGON ((-0.5 -0.5, -0.5 0.5, 0.5 0.5, 0.5 -0.5, -0.5 -0.5))`
...
+Since: `v1.2.1`
+
+Examples:
+
+```SQL
+SELECT ST_ExteriorRing(df.geometry)
+FROM df
+```
+
+Input: `POLYGON((0 0, 1 1, 2 1, 0 1, 1 -1))`
```
Comment (jiayuasu) [inline docs/api/flink/Function.md:198]: The example here is NOT a valid WKT format

File: flink/src/test/java/org/apache/sedona/flink/FunctionTest.java [hunk 5]
```diff
@@ -82,6 +87,27 @@ public void testReverse() {
...
+    public void testPointN_negativeN() {
+        int n = -3;
+        Table polygonTable = createPolygonTable(1);
+        Table linestringTable = polygonTable.select(call(Functions.ST_ExteriorRing.class.getSimpleName(), $(polygonColNames[0])));
+        Table pointTable = linestringTable.select(call(Functions.ST_PointN.class.getSimpleName(), $("_c0"), n));
+        Point point = (Point) first(pointTable).getField(0);
+        assert point != null;
+        Assert.assertEquals("POINT (0.5 0.5)", point.toString());
+    }
+
```
Comment (jiayuasu) [inline flink/src/test/java/org/apache/sedona/flink/FunctionTest.java:110]: Please add a test for exterior ring.

Comment (Pooja444) [reply flink/src/test/java/org/apache/sedona/flink/FunctionTest.java:111]: Added

File: docs/api/flink/Function.md (hunk 4 above)
Comment (Pooja444) [reply docs/api/flink/Function.md]: Updated it

File: sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Functions.scala [hunk 6]
```diff
@@ -1620,6 +1620,47 @@ case class ST_Reverse(inputExpressions: Seq[Expression])
...
+ * @param inputExpressions sequence of 2 input arguments, a geometry and a value 'n'
+ */
+case class ST_PointN(inputExpressions: Seq[Expression])
+  extends Expression with CodegenFallback {
+  inputExpressions.validateLength(2)
+
+  override def nullable: Boolean = true
+
+  lazy val GeometryFactory = new GeometryFactory()
+  lazy val emptyGeometry: GeometryCollection = GeometryFactory.createGeometryCollection(null)
```
Comment (jiayuasu) [inline sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Functions.scala:1634]: This emptyGeometry is NOT consistent with PostGIS requirement. ST_PointN should return NULL if no input linestring. You return an empty GeometryCollection but not NULL. https://postgis.net/docs/ST_PointN.html

Please fix this and add corresponding test cases for non-line string input.

**Review by jiayuasu — CHANGES_REQUESTED** [review]

**Review by jiayuasu — APPROVED** [review]

**Pull request merged**